[FEAT] Add ROS compatible zenoh sink

kamibo opened the pull request:
This PR adds a ROS compatible Zenoh sink to forward data from Copper applications to ROS 2 via the Zenoh middleware (rmw_zenoh).

- Introduces a new sink (cu-zenoh-ros-sink) with full implementation including liveliness, key expression formatting, and configuration handling.
- Adds an example application (cu_zenoh_ros) and updates workspace members and dependencies accordingly.

makeecat reviewed (changes requested):
This PR adds a new zenoh sink component that allows Copper to publish data to ROS 2 using the Zenoh middleware. This enables seamless integration between Copper applications and ROS 2 nodes, using the RMW zenoh implementation.

Several Actions items:
1. Consistency of Error Msg: Please either use `ZenohRosSink` or `ZenohSink` as the prefix of error msg.
2. Add more documentation for public API
3. Fix `zenoh` version back to `1.3.4` to avoid surprise
4. The CDR serialization adds an additional layer of conversion that could impact performance, worth noting it in README for future improvements.
5. In `components/sinks/cu_zenoh_ros_sink/src/lib.rs` (process method), CDR serialization uses `.unwrap()` which can panic, please handle the error:
  `let serial_ros_payload = cdr::serialize::<_, _, CdrBe>(&ros_payload, Infinite).unwrap();`

kamibo commented:
> This PR adds a new zenoh sink component that allows Copper to publish data to ROS 2 using the Zenoh middleware. This enables seamless integration between Copper applications and ROS 2 nodes, using the RMW zenoh implementation.
> 
> Several Actions items:
> 
>     1. Consistency of Error Msg: Please either use `ZenohRosSink` or `ZenohSink` as the prefix of error msg.
> 
>     2. Add more documentation for public API
> 
>     3. Fix `zenoh` version back to `1.3.4` to avoid surprise
> 
>     4. The CDR serialization adds an additional layer of conversion that could impact performance, worth noting it in README for future improvements.
> 
>     5. In `components/sinks/cu_zenoh_ros_sink/src/lib.rs` (process method), CDR serialization uses `.unwrap()` which can panic, please handle the error:
>        `let serial_ros_payload = cdr::serialize::<_, _, CdrBe>(&ros_payload, Infinite).unwrap();`

All done.
Regarding the API documentation, I refactored the ROS message adapter trait to make it easier to use and added some comments.
Let me know if there’s anything else we can do to improve it.

makeecat reviewed (approved):
Works for me. @gbin Do you have some comments on the use of string ? I believe we can improve that in a second PR to have a baseline.

kamibo commented:
TIL: They already planned to change the attachment format for the next release https://github.com/ros2/rmw_zenoh/issues/600  :disappointed:

gbin commented on components/payloads/cu_ros_payloads/src/builtin.rs:
```diff
@@ -0,0 +1,13 @@
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Header {
+    pub stamp: Time,
+    pub frame_id: std::string::String,
```
Nit: if the string is short maybe use a short string like compact_str we use in cu29_runtime?

gbin commented on components/payloads/cu_ros_payloads/src/lib.rs:
```diff
@@ -0,0 +1,17 @@
+    fn namespace() -> &'static str;
+    fn type_name() -> &'static str;
+    fn convert(&self) -> Self::Output;
+}
```
I am not sure about forcing static here, if you parametrize the type I think the implementor/users of that trait can always use static for it:

```Rust
pub trait RosMsgAdapter<'a> {
    type Output: Serialize;

    fn namespace() -> &'a str;
    fn type_name() -> &'a str;
    fn convert(&self) -> Self::Output;
}
``` 

```Rust
impl<'a> RosMsgAdapter<'a> for MyMsg {
    type Output = MyRosCompatibleMsg;

    fn namespace() -> &'a str {
        "std_msgs"
    }

    fn type_name() -> &'a str {
        "String"
    }

    fn convert(&self) -> Self::Output {
        MyRosCompatibleMsg { ... }
    }
}
```

kamibo replied:
Let me known if it sounds better now.

gbin replied:
I was thinking about the convert method ... this is just a not very idiomatic way of a From/Into, you can force the trait with a + 

```
pub trait RosMsgAdapter<'a>: Into<Self::Output> + Sized
where
    Self::Output: Serialize + 'a
{
    fn namespace() -> &'a str;
    fn type_name() -> &'a str {
        ros_type_name!(Self::Output)
    }
}
```
^ that feels more like the signature I would expect

gbin commented on components/payloads/cu_ros_payloads/Cargo.toml:
```diff
@@ -0,0 +1,16 @@
+categories.workspace = true
+homepage.workspace = true
+repository.workspace = true
+description = "Those are ROS compatible payloads"
```
This will be the basic public description on Crates.io, maybe mention that it is part of copper/extra payload for copper.

gbin commented on components/payloads/cu_ros_payloads/src/sensor_msgs.rs:
```diff
@@ -0,0 +1,54 @@
+// sensor_msgs/PointField
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PointField {
+    pub name: String,
```
maybe small_str

gbin commented on components/payloads/cu_ros_payloads/src/sensor_msgs.rs:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    #[allow(unreachable_code)]
+    fn convert(&self) -> Self::Output {
```
can't we just use the standard "Into" facility of Rust here?

gbin commented on components/payloads/cu_ros_payloads/src/std_msgs.rs:
```diff
@@ -0,0 +1,25 @@
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Int8 {
+    pub data: i8,
+}
```
maybe Int8(i8)?

kamibo replied:
It should work but I think it is better to stick to ROS implementation https://docs.ros.org/en/melodic/api/std_msgs/html/msg/Int8.html

gbin commented on components/payloads/cu_ros_payloads/src/std_msgs.rs:
```diff
@@ -0,0 +1,25 @@
+        "std_msgs"
+    }
+
+    fn type_name() -> &'static str {
```
if the type_name is almost always the Struct name you can provide it for the user with a small default impl in the trait maybe? 

```Rust
pub trait RosMsgAdapter<'a> {
    type Output: serde::Serialize;

    fn convert(&self) -> Self::Output;

    fn namespace() -> &'a str;

    fn type_name() -> &'a str {
        ros_type_name!(Self::Output)
    }
}

#[macro_export]
macro_rules! ros_type_name {
    ($t:ty) => {{
        std::any::type_name::<$t>()
            .rsplit("::")
            .next()
            .unwrap()
    }};
}

```

link to the playground with a little test:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=c10055c62bf8aadd4e77e9920d3965a7

note that the user can override this behavior in the implementation

gbin commented on components/sinks/cu_zenoh_ros_sink/README.md:
```diff
@@ -0,0 +1,66 @@
+        (
+            id: "publisher",
+            config: {
+                "zenoh_config_file": "/opt/ros/humble/share/rmw_zenoh_cpp/config/DEFAULT_RMW_ZENOH_SESSION_CONFIG.json5",
```
That breaks the "self-contained" nature of the Copper config.

We do an include! of the RON file at compile time to be able to have a 1 file deployment and a path to a no_std deployment too: ie. *no* filesystem at all.

is it possible, maybe in another PR to provide an embedded version of it: for example a config_str: and the inline version of the json file?

gbin commented on components/sinks/cu_zenoh_ros_sink/src/error.rs:
```diff
@@ -0,0 +1,10 @@
+use cu29::CuError;
+use zenoh::Error as ZenohError;
+
+pub fn cu_error(msg: &str, error: ZenohError) -> CuError {
```
Ok those ones are getting ridiculous, I am adding a note to refactor CuError with something like:
```Rust
#[derive(Debug)]
pub struct CuError {
    message: String,
    cause: Option<Box<dyn std::error::Error + Send + Sync>>,
}
```
No action needed on your part yet :)
https://github.com/copper-project/copper-rs/issues/317

gbin commented on components/sinks/cu_zenoh_ros_sink/src/lib.rs:
```diff
@@ -0,0 +1,190 @@
+
+pub struct ZenohRosContext {
+    session: zenoh::Session,
+    publisher: zenoh::pubsub::Publisher<'static>,
```
The publisher has to be static here?

kamibo replied:
Actually, as far as I understand, it only means that the "key expr" is owned. But maybe I'm missing something.

gbin replied:
So what I understand is that basically either it is a static string OR you won't be able to deallocate the publisher (Box::leak), so the ZenohRosContext, so the session etc.. ?

I believe you need to forward those lifetimes to the Context.

'static is really what it means, it will be never ever be deallocated.

kamibo replied:
> I believe you need to forward those lifetimes to the Context.

I also tried this but I can't figure out to forward those lifetimes beyond code generated from the RON file.
But I'm not sure it is a big deal.
This lifetime is related to the keyexpr stored inside the publisher and according the documentation
https://docs.rs/zenoh-keyexpr/latest/zenoh_keyexpr/

> [KeyExpr](https://docs.rs/zenoh/latest/zenoh/key_expr/struct.KeyExpr.html) works like a [Cow<str>](https://doc.rust-lang.org/nightly/alloc/borrow/enum.Cow.html), but also stores some additional context internal to  Zenoh to optimize routing and network usage.

AFIU this means that using static lifetime implies key expr cannot be borrowed https://github.com/eclipse-zenoh/zenoh/blob/63a6e33fe58eb98485c13dad0924f673baedc060/zenoh/src/api/key_expr.rs#L40
And we don't care about that as we have to compute the key expr.

> 'static is really what it means, it will be never ever be deallocated.

Not necessarily. My understanding is it can be hold for as long as needed.

https://doc.rust-lang.org/rust-by-example/scope/lifetime/static_lifetime.html#trait-bound

But maybe I'm missing something...

gbin commented on components/sinks/cu_zenoh_ros_sink/src/lib.rs:
```diff
@@ -0,0 +1,190 @@
+
+impl<'cl, P> CuSinkTask<'cl> for ZenohRosSink<P>
+where
+    P: CuMsgPayload + RosMsgAdapter + 'cl + 'static,
```
while correct rust, this 'static feels out of place. Conceptually those types only need to live for the duration of one CopperList ie. one execution cycle.

gbin reviewed (changes requested):
Awesome PR,

tried to prod on some things what looked weird, feel free to push back.

Thank you so much!

kamibo commented:
Lots of good points :+1:
Note: I'm going be busy for a week (team meeting in Paris).  I'll continue when I get back

edgarriba commented on components/sinks/cu_zenoh_ros_sink/src/lib.rs:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    fn start(&mut self, _clock: &RobotClock) -> CuResult<()> {
+        let session = zenoh::Wait::wait(zenoh::open(self.config.session.clone()))
```
How heavy it’s to create a new session per task, instead of having somehow a global ?

edgarriba replied:
in a separated projet using copper I have instantiated within a lazy operator

kamibo replied:
This sounds great. However I'm wondering how to define a single zenoh session config for all the tasks inside the RON file? :thinking: 
@gbin any thoughts about this?

gbin replied:
I think we really need a new kind of "service" task just holding a state / running a daemon thread. I propose to add that as an issue for that and discuss the arch and determinism on the issue. For now let's move forward with this PR with the known limitation then refactor it with the new facility in the framework.

edgarriba replied:
really good idea. Looking forward to it

gbin commented:
Ok let's merge this and go back to it for improvements, I don't want the PR to go stale.

Thanks everyone
